### PR TITLE
Sync: Favicons Fetching

### PR DESCRIPTION
**Pull request by malmstein**
Task/Issue URL: https://app.asana.com/0/488551667048375/1206440765317542/f

### Steps to test this PR
Easier to test with two devices.
We’ll assume that device A has some bookmarks / favourites and tests are done in device B
You can use https://github.com/duckduckgo/sync_crypto/actions/workflows/create-test-account.yml to create a staging test account

_Bottom Sheet (Bookmarks)_
- [ ] Enable Sync and join account from device A
- [ ] Open bookmarks screen
- [ ] Verify Bottom Sheet appears
- [ ] Enable Favicons Fetching
- [ ] Verify favicons appear
- [ ] Close screen and open it again
- [ ] Verify Bottom Sheet does not appear

_Bottom Sheet (Bookmarks)_
- [ ] Enable Sync and join account from device A
- [ ] Open bookmarks screen
- [ ] Verify Bottom Sheet appears
- [ ] Dismiss sheet
- [ ] Verify favicons do not appear
- [ ] Close screen and open it again
- [ ] Verify Bottom Sheet does not appear

_Bottom Sheet (New Tab)_
- [ ] Enable Sync and join account from device A
- [ ] Open New Tab
- [ ] Verify Bottom Sheet appears
- [ ] Enable Favicons Fetching
- [ ] Verify favicons appear

_Bottom Sheet (New Tab)_
- [ ] Enable Sync and join account from device A
- [ ] Open New Tab
- [ ] Verify Bottom Sheet appears
- [ ] Dismiss sheet
- [ ] Verify favicons do not appear

_Enable from Sync Settings_
- [ ] Enable Sync and join account from device A
- [ ] Enable Favicons Fetching
- [ ] Open Bookmarks
- [ ] Verify favicons appear
- [ ] Open Favourites
- [ ] Verify favicons appear

_Enable from Sync Settings - Bottom Sheet doesn’t appear_
- [ ] Fresh install
- [ ] Enable Sync and join account from device A
- [ ] Enable Favicons Fetching
- [ ] Open Bookmarks
- [ ] Verify favicons appear
- [ ] Verify bottom sheet doesn’t appear
- [ ] Open Sync Settings and Disable Favicons Fetching
- [ ] Open bookmarks screen
- [ ] Verify Bottom Sheet appears

_Favicons fetching logic_
Easier tested with the test account
- [ ] Fresh install
- [ ] Join the test account
- [ ] Enable favicons fetching
- [ ] Open bookmarks and verify favicons appear
- [ ] Open sports folder and verify NFL favicon doesn’t appear
- [ ] Open NFL site and wait for it to load (this trigger existing favicon fetching logic)
- [ ] Open bookmarks screen and sports folder
- [ ] Verify NFL favicon appears

### CODE REVIEW

**Comment by malmstein** [comment]
* `develop`
  * **#4124** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4124?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @malmstein and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt [hunk 1]
```diff
@@ -466,6 +491,7 @@ class BookmarksActivity : DuckDuckGoActivity() {
 
     companion object {
         const val SAVED_SITE_URL_EXTRA = "SAVED_SITE_URL_EXTRA"
+        private const val TAG_FAVICON_PROMPT_DIALOG = "TAG_FAVICON_PROMPT_DIALOG"
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt:494]: unused

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt [hunk 2]
```diff
@@ -339,4 +353,26 @@ class BookmarksViewModel @Inject constructor(
...
+    ) {
+        viewModelScope.launch(dispatcherProvider.io()) {
+            faviconsFetchingPrompt.onPromptAnswered(fetchingEnabled)
+        }
+        if (fetchingEnabled) {
+            refreshBookmarks(currentFolderId)
+        }
+    }
+
+    private fun refreshBookmarks(currentFolderId: String) {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksViewModel.kt:369]: this method/intention is not clear. It seems we want to refresh bookmarks, but I think this is somehow related to the favicons.

I see we call it after fetchingEnabled. My assumption is to force some redraw. Instead why not a command that is more explicit about the action we want to perform. And then you do in the adapter what is required? that way the intention is more clear, and the required logic close to the component that renders. Otherwise it's not clear why we are refreshing.

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt [hunk 3]
```diff
@@ -176,20 +188,33 @@ class BookmarksActivity : DuckDuckGoActivity() {
...
+                is DeleteBookmarkFolder -> deleteBookmarkFolder(it.bookmarkFolder)
+                is ConfirmDeleteBookmarkFolder -> confirmDeleteBookmarkFolder(it.bookmarkFolder)
+                is LaunchBookmarkImport -> launchBookmarkImport()
+                is ShowFaviconsPrompt -> showFaviconsPrompt()
             }
         }
     }
 
+    private fun showFaviconsPrompt() {
+        val faviconPrompt = FaviconPromptSheet.Builder(this)
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt:207]: Wondering if this could be a custom view with its own viewmodel? similar to how we are implementing some view like NotifymeView, or the view we show in syncsettings.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 4]
```diff
@@ -1067,10 +1068,11 @@ class BrowserTabFragment :
     }
 
     private fun showHome() {
-        viewModel.clearPreviousAppLink()
+        viewModel.onHomeShown()
         dismissAppLinkSnackBar()
         errorSnackbar.dismiss()
         newBrowserTab.newTabLayout.show()
+        // show the favicons bottom sheet if needed
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:1075]: leftover? I dont' see any ref to bottom sheet here.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 5]
```diff
@@ -539,6 +542,8 @@ class BrowserTabViewModel @Inject constructor(
         data class WebShareRequest(val data: JsCallbackData) : Command()
         data class ScreenLock(val data: JsCallbackData) : Command()
         object ScreenUnlock : Command()
+
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:545]: (remove new line)

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 6]
```diff
@@ -3233,6 +3273,26 @@ class BrowserTabViewModel @Inject constructor(
         }
     }
 
+    fun onHomeShown() {
+        Timber.d("Sync: onHomeShown, checking for favicons prompt")
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:3277]: 👀  "Sync"?

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 7]
```diff
@@ -3744,6 +3759,7 @@ class BrowserTabFragment :
             configuration: HomePanelCta,
             favorites: List<QuickAccessFavorite>,
         ) {
+            Timber.d("Sync: showHomeCta $favorites")
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:3762]: 👀

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:1]: As discussed and as feedback, all those code format changes add noise here. I will move the discussion to the weekly and see how we can avoid this.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt [hunk 8]
```diff
@@ -3233,6 +3273,26 @@ class BrowserTabViewModel @Inject constructor(
         }
     }
 
+    fun onHomeShown() {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabViewModel.kt:3276]: should be tested?

File: app/src/main/java/com/duckduckgo/app/browser/favicon/DuckDuckGoFaviconManager.kt [hunk 9]
```diff
@@ -93,6 +95,20 @@ class DuckDuckGoFaviconManager constructor(
         }
     }
 
+    override suspend fun tryFetchFaviconForUrl(url: String): File? {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/favicon/DuckDuckGoFaviconManager.kt:98]: could this be private?

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/favicons/FaviconsFetchingStore.kt
Comment (cmonfortep) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/favicons/FaviconsFetchingStore.kt:1]: I believe we discussed this being part of `saved-sites` module, not `sync`. Let me know if you want to discuss about it.

File: sync/sync-api/src/main/java/com/duckduckgo/sync/api/favicons/FaviconsFetchingPrompt.kt
Comment (cmonfortep) [inline sync/sync-api/src/main/java/com/duckduckgo/sync/api/favicons/FaviconsFetchingPrompt.kt:1]: ditto. This should be in savedsites module.

File: sync/sync-impl/src/main/res/values/donottranslate.xml [hunk 10]
```diff
@@ -38,9 +38,19 @@
...
     <string name="sync_internal_account_header">Account Settings</string>
+    <string name="sync_internal_device_header">Device Settings</string>
     <string name="sync_internal_recovery_code">Recovery code</string>
     <string name="sync_internal_recovery_code_cta">Copy Recovery code</string>
+    <string name="sync_internal_favicons_prompt_cta">Reset Favicons Prompt</string>
     <string name="userIdSectionHeader">User ID</string>
     <string name="deviceIdSectionHeader">Device Id</string>
     <string name="deviceNameSectionHeader">Device Name</string>
+
+    <!-- Sync Favicons -->
```
Comment (cmonfortep) [inline sync/sync-impl/src/main/res/values/donottranslate.xml:49]: All these screens should be in saved-sites module.

File: app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchActivity.kt [hunk 11]
```diff
@@ -474,6 +474,10 @@ class SystemSearchActivity : DuckDuckGoActivity() {
         return intent.action == NEW_SEARCH_ACTION
     }
 
+    private fun launchedFromSystemVoiceSearchBox(intent: Intent): Boolean {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/systemsearch/SystemSearchActivity.kt:477]: Is this related to the PR?

File: app/src/main/res/layout/view_sync_favicons_fetching.xml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+             android:layout_width="match_parent"
+             android:layout_height="wrap_content">
+
+    <com.duckduckgo.common.ui.view.listitem.TwoLineListItem
+        android:id="@+id/sync_favicons_fetching"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:leadingIcon="@drawable/ic_image_24"
+        app:leadingIconBackground="circular"
+        app:primaryText="@string/sync_fetch_icons_title"
```
Comment (cmonfortep) [inline app/src/main/res/layout/view_sync_favicons_fetching.xml:29]: We shouldn't access string from other modules.

File: app/src/main/res/layout/bottom_sheet_favicons_prompt.xml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                app:layout_constraintStart_toStartOf="@id/faviconsPromptPrimaryCta"
+                app:layout_constraintTop_toBottomOf="@id/faviconsPromptTitle" />
+
+            <com.duckduckgo.common.ui.view.button.DaxButtonPrimary
+                android:id="@+id/faviconsPromptPrimaryCta"
+                android:layout_width="0dp"
+                app:buttonSize="large"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="@dimen/keyline_5"
+                android:text="@string/faviconsPromptPrimaryCta"
```
Comment (cmonfortep) [inline app/src/main/res/layout/bottom_sheet_favicons_prompt.xml:95]: ditto, we shouldn't be accessing strings or any other resource from another module

**Review by cmonfortep — COMMENTED** [review]
I've only done code review here. Sending feedback first because I see some blocking changes here.

Main topics:
- Saved sites module should own favicon fetcher and any logic related. We shouldn't be adding any new interface to sync module
- We should avoid accessing resources from other modules (e.g: app should not access resources from sync)

File: app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt (hunk 3 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/bookmarks/ui/BookmarksActivity.kt:207]: I thought about it, but we have the issue that this is not a `View` attached to the tree hierarchy until it’s visible because it’s a `BottomSheet`. Meaning, we can’t inject `ViewModels` using the `findViewTreeViewModelStoreOwner`.

Since this is only used in two places, I found it easier to treat it as a simple bottom sheet, and let the `ViewModels` of the screens where it’s shown to deal with the logic.

File: app/src/main/java/com/duckduckgo/app/browser/favicon/DuckDuckGoFaviconManager.kt (hunk 9 above)
Comment (malmstein) [reply app/src/main/java/com/duckduckgo/app/browser/favicon/DuckDuckGoFaviconManager.kt:98]: No, this is used to prefetch favicons from the browser and update the tab. It’s not a new addition

**Review by cmonfortep — APPROVED** [review]
LGTM 

Will create follow up task in Asana as per discussion
